File: go/exercises/practice/unit_test_12e9343a_954a_4b59_ae9c_341d55f41456/multi_flow/.meta/multi_flow.go
package multiflow

// MaxMultiFlow calculates the maximum flow for multiple commodities in a network
func MaxMultiFlow(n int, edges [][3]int, commodities [][3]int) []int {
    // Create result array for each commodity
    flows := make([]int, len(commodities))
    
    // Create residual graph capacity matrix
    residual := make([][]int, n)
    for i := range residual {
        residual[i] = make([]int, n)
    }
    
    // Initialize residual graph with edge capacities
    for _, edge := range edges {
        residual[edge[0]][edge[1]] = edge[2]
    }
    
    // Process each commodity independently
    for i, commodity := range commodities {
        source := commodity[0]
        sink := commodity[1]
        demand := commodity[2]
        
        // Create a temporary residual graph for this commodity
        tempResidual := make([][]int, n)
        for j := range tempResidual {
            tempResidual[j] = make([]int, n)
            copy(tempResidual[j], residual[j])
        }
        
        flows[i] = findMaxFlow(n, tempResidual, source, sink, demand)
        
        // Update the main residual graph with the flow used by this commodity
        for j := 0; j < n; j++ {
            for k := 0; k < n; k++ {
                residual[j][k] = tempResidual[j][k]
            }
        }
    }
    
    return flows
}

// findMaxFlow implements the Edmonds-Karp algorithm to find the maximum flow
func findMaxFlow(n int, residual [][]int, source, sink, demand int) int {
    totalFlow := 0
    
    for {
        // Find augmenting path using BFS
        parent := make([]int, n)
        for i := range parent {
            parent[i] = -1
        }
        
        // BFS queue
        queue := []int{source}
        parent[source] = source
        
        for len(queue) > 0 && parent[sink] == -1 {
            current := queue[0]
            queue = queue[1:]
            
            for next := 0; next < n; next++ {
                if parent[next] == -1 && residual[current][next] > 0 {
                    parent[next] = current
                    queue = append(queue, next)
                }
            }
        }
        
        // If no augmenting path is found, break
        if parent[sink] == -1 {
            break
        }
        
        // Find minimum residual capacity along the path
        pathFlow := demand - totalFlow
        current := sink
        for current != source {
            prev := parent[current]
            if residual[prev][current] < pathFlow {
                pathFlow = residual[prev][current]
            }
            current = prev
        }
        
        // Update residual capacities
        current = sink
        for current != source {
            prev := parent[current]
            residual[prev][current] -= pathFlow
            residual[current][prev] += pathFlow
            current = prev
        }
        
        totalFlow += pathFlow
        
        // If we've met the demand, break
        if totalFlow >= demand {
            return demand
        }
    }
    
    return totalFlow
}

File: python/exercises/practice/unit_test_a4ad7b2a_1889_4b2b_a841_bda5f8be9da2/multi_flow/.meta/multi_flow.py
import math
from collections import defaultdict, deque

class Edge:
    def __init__(self, u, v, capacity, cost, rev, orig, orig_key):
        self.u = u            # tail node
        self.v = v            # head node
        self.capacity = capacity
        self.cost = cost
        self.rev = rev        # index of reverse edge in graph[v]
        self.orig = orig      # True if this is an original (forward) edge
        self.orig_key = orig_key  # (u, v) tuple if orig==True, else None

def multi_flow(num_nodes, edges, commodities):
    # Build the residual graph
    graph = [[] for _ in range(num_nodes)]
    # For each input edge, add forward and reverse edges.
    for u, v, cap, cost in edges:
        # Forward edge
        forward = Edge(u, v, cap, cost, len(graph[v]), True, (u, v))
        # Reverse edge
        reverse = Edge(v, u, 0, -cost, len(graph[u]), False, None)
        graph[u].append(forward)
        graph[v].append(reverse)

    # Global result: dictionary mapping (u, v) -> {commodity_index: flow}
    result = defaultdict(lambda: defaultdict(int))

    # For each commodity, use a successive shortest path algorithm on the residual graph.
    for comm_index, (source, sink, demand) in enumerate(commodities):
        flow_needed = demand
        # While there is remaining flow for this commodity.
        while flow_needed > 0:
            # Bellman-Ford to find the shortest path from source to sink.
            dist = [math.inf] * num_nodes
            in_queue = [False] * num_nodes
            prev_node = [-1] * num_nodes
            prev_edge = [-1] * num_nodes
            dist[source] = 0

            queue = deque([source])
            in_queue[source] = True

            while queue:
                u = queue.popleft()
                in_queue[u] = False
                for i, e in enumerate(graph[u]):
                    if e.capacity > 0 and dist[e.v] > dist[u] + e.cost:
                        dist[e.v] = dist[u] + e.cost
                        prev_node[e.v] = u
                        prev_edge[e.v] = i
                        if not in_queue[e.v]:
                            queue.append(e.v)
                            in_queue[e.v] = True

            # If sink is unreachable, then the commodity demand cannot be satisfied.
            if dist[sink] == math.inf:
                return None

            # Determine the maximum flow we can send along the found path.
            path_flow = flow_needed
            v = sink
            while v != source:
                u = prev_node[v]
                edge = graph[u][prev_edge[v]]
                if edge.capacity < path_flow:
                    path_flow = edge.capacity
                v = u

            # Augment the flow along the path.
            v = sink
            while v != source:
                u = prev_node[v]
                edge = graph[u][prev_edge[v]]
                # If this is an original forward edge, record the commodity flow.
                if edge.orig:
                    result[edge.orig_key][comm_index] += path_flow
                # Decrease capacity and increase capacity of the reverse edge.
                edge.capacity -= path_flow
                rev_edge = graph[edge.v][edge.rev]
                rev_edge.capacity += path_flow
                v = u

            flow_needed -= path_flow

    # Convert result to regular dicts.
    final_result = {}
    for key, comm_dict in result.items():
        final_result[key] = dict(comm_dict)
    return final_result

if __name__ == '__main__':
    # For simple ad-hoc testing
    num_nodes = 4
    edges = [
        (0, 1, 10, 1),
        (0, 2, 5, 2),
        (1, 2, 15, 1),
        (1, 3, 7, 3),
        (2, 3, 12, 1)
    ]
    commodities = [
        (0, 3, 8)
    ]
    solution = multi_flow(num_nodes, edges, commodities)
    if solution is None:
        print("No feasible solution")
    else:
        print(solution)